kuma/core/utils.py

```diff
@@ -539,3 +542,63 @@ def safer_pyquery(*args, **kwargs):
         args = (f" {args[0]}",) + args[1:]
 
     return pq(*args, **kwargs)
+
+
+def send_mail_retrying(
+    subject,
+    message,
+    from_email,
+    recipient_list,
+    fail_silently=False,
+    auth_user=None,
+    auth_password=None,
+    connection=None,
+    html_message=None,
+    **kwargs,
+):
+    """Copied verbatim from django.core.mail.send_mail but with the override
+    that we're using our EmailMultiAlternativesRetrying class instead.
+    See its doc string for its full documentation.
+
+    The only difference is that this function allows for setting your
+    own custom 'retrying' keyword argument.
+    """
+    connection = connection or get_connection(
+        username=auth_user, password=auth_password, fail_silently=fail_silently,
+    )
+    mail = EmailMultiAlternativesRetrying(
+        subject, message, from_email, recipient_list, connection=connection
+    )
+    if html_message:
+        mail.attach_alternative(html_message, "text/html")
+
+    return mail.send(**kwargs)
+
+
+class EmailMultiAlternativesRetrying(EmailMultiAlternatives):
+    """
+    Thin wrapper on django.core.mail.EmailMultiAlternatives that adds
+    a retrying functionality. By default, the only override is that
+    we're very explicit about the of exceptions we treat as retry'able.
+    The list of exceptions we use to trigger a retry are:
+
+        * smtplib.SMTPServerDisconnected
+
+    Only list exceptions that have been known to happen and are safe.
+    """
+
+    def send(self, *args, **kwargs):
+        # See https://github.com/mozilla-releng/redo
+        # for a list of the default options to the redo.retry function
+        # which the redo.retrying context manager wraps.
+        retry_options = kwargs.pop(
```

Comment:
Might be a "I'm missing python subtleties"-kind of question: But why pop kwargs here and just specify retrying in 590 explicitly?

Reply:
Normally, when you override a method and want to add some extra keyword arguments you do something like:
```python
def send(self, *args, special="for this override", **kwargs):
```
that way you're not messing with `kwargs` with keys that are only relevant in the override. 
But the default this time is so much more complex (unlike the above example `"for this override"`). 

I guess it could be something like this instead:
```python
def send(self, *args, retrying=None, **kwargs):
    if retrying is None:
        retrying = { "retry_exceptions": .... }
```
Do you think that's prettier?

Reply:
I think I do like that more, and I'd be even fine with having the default be that dict. Though a pitfall in Python is that those defaults are only initialied once, right? So all invocations of that functions would have shared mutable access to the same object, which might be too dangerous here?